File: backend/schemas/kanban.py
#
# 项目管理看板Pydantic模型
# 用于API请求/响应的数据验证
#

from pydantic import BaseModel, Field, validator
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum

# 枚举定义
class PriorityLevel(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    URGENT = "urgent"

class CardStatus(str, Enum):
    TODO = "todo"
    IN_PROGRESS = "in_progress"
    DONE = "done"
    ARCHIVED = "archived"

# ===== Board 模型 =====
class BoardBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255, description="看板名称")
    description: Optional[str] = Field(None, max_length=1000, description="看板描述")
    background_color: Optional[str] = Field("#FFFFFF", pattern=r"^#[0-9A-Fa-f]{6}$", description="背景颜色")
    background_image: Optional[str] = Field(None, max_length=500, description="背景图片URL")

class BoardCreate(BoardBase):
    """创建看板请求模型"""
    pass

class BoardUpdate(BaseModel):
    """更新看板请求模型"""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=1000)
    background_color: Optional[str] = Field(None, pattern=r"^#[0-9A-Fa-f]{6}$")
    background_image: Optional[str] = Field(None, max_length=500)
    is_archived: Optional[bool] = None

class BoardResponse(BoardBase):
    """看板响应模型"""
    id: str
    user_id: str
    is_archived: bool
    position: int
    created_at: datetime
    updated_at: datetime
    lists: List['ListResponse'] = []

    class Config:
        orm_mode = True

# ===== List 模型 =====
class ListBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255, description="列表名称")
    description: Optional[str] = Field(None, max_length=1000, description="列表描述")
    position: int = Field(0, ge=0, description="在面板中的位置")

class ListCreate(ListBase):
    """创建列表请求模型"""
    board_id: str = Field(..., description="所属看板ID")

class ListUpdate(BaseModel):
    """更新列表请求模型"""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=1000)
    position: Optional[int] = Field(None, ge=0)
    is_archived: Optional[bool] = None

class ListResponse(ListBase):
    """列表响应模型"""
    id: str
    board_id: str
    is_archived: bool
    created_at: datetime
    updated_at: datetime
    cards: List['CardResponse'] = []

    class Config:
        orm_mode = True

# ===== Card 模型 =====
class CardBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=500, description="卡片标题")
    description: Optional[str] = Field(None, description="卡片描述，支持Markdown")
    position: int = Field(0, ge=0, description="在列表中的位置")
    due_date: Optional[datetime] = Field(None, description="截止日期")
    priority: PriorityLevel = Field(PriorityLevel.MEDIUM, description="优先级")
    color: Optional[str] = Field(None, pattern=r"^#[0-9A-Fa-f]{6}$", description="卡片颜色")
    tags: List[str] = Field(default_factory=list, description="标签列表")

class CardCreate(CardBase):
    """创建卡片请求模型"""
    list_id: str = Field(..., description="所属列表ID")

class CardUpdate(BaseModel):
    """更新卡片请求模型"""
    title: Optional[str] = Field(None, min_length=1, max_length=500)
    description: Optional[str] = Field(None)
    position: Optional[int] = Field(None, ge=0)
    due_date: Optional[datetime] = Field(None)
    priority: Optional[PriorityLevel] = None
    is_completed: Optional[bool] = None
    color: Optional[str] = Field(None, pattern=r"^#[0-9A-Fa-f]{6}$")
    tags: Optional[List[str]] = None

class CardMoveRequest(BaseModel):
    """卡片移动请求模型"""
    source_list_id: str = Field(..., description="源列表ID")
    target_list_id: str = Field(..., description="目标列表ID")
    new_position: int = Field(..., ge=0, description="新位置")

class CardReorderRequest(BaseModel):
    """卡片重新排序请求模型"""
    new_position: int = Field(..., ge=0, description="新位置")

class CardResponse(CardBase):
    """卡片响应模型"""
    id: str
    list_id: str
    is_completed: bool
    is_archived: bool
    created_at: datetime
    updated_at: datetime
    completed_at: Optional[datetime]

    class Config:
        orm_mode = True

# ===== 看板统计模型 =====
class BoardStats(BaseModel):
    """看板统计信息"""
    total_cards: int
    completed_cards: int
    overdue_cards: int
    cards_by_priority: Dict[str, int]
    cards_by_list: Dict[str, int]

# ===== 批量操作模型 =====
class BulkCardUpdate(BaseModel):
    """批量更新卡片请求"""
    card_ids: List[str] = Field(..., min_items=1, description="要更新的卡片ID列表")
    updates: CardUpdate = Field(..., description="要应用的更新")

class BulkCardMove(BaseModel):
    """批量移动卡片请求"""
    card_ids: List[str] = Field(..., min_items=1, description="要移动的卡片ID列表")
    target_list_id: str = Field(..., description="目标列表ID")

# ===== 搜索和过滤模型 =====
class CardFilter(BaseModel):
    """卡片过滤条件"""
    list_ids: Optional[List[str]] = Field(None, description="列表ID过滤")
    priorities: Optional[List[PriorityLevel]] = Field(None, description="优先级过滤")
    is_completed: Optional[bool] = Field(None, description="完成状态过滤")
    has_due_date: Optional[bool] = Field(None, description="是否有截止日期")
    is_overdue: Optional[bool] = Field(None, description="是否已过期")
    tags: Optional[List[str]] = Field(None, description="标签过滤")

class SearchRequest(BaseModel):
    """搜索请求模型"""
    query: str = Field(..., min_length=1, description="搜索关键词")
    board_id: Optional[str] = Field(None, description="限定看板范围")
    include_archived: bool = Field(False, description="是否包含已归档内容")

class SearchResponse(BaseModel):
    """搜索响应模型"""
    cards: List[CardResponse]
    total_count: int
    query: str

# 解决前向引用
BoardResponse.update_forward_refs()
ListResponse.update_forward_refs()

File: frontend/src/types/kanban.ts
/**
 * 项目管理看板类型定义
 */

// 枚举定义
export enum PriorityLevel {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  URGENT = 'urgent'
}

export enum CardStatus {
  TODO = 'todo',
  IN_PROGRESS = 'in_progress',
  DONE = 'done',
  ARCHIVED = 'archived'
}

// ===== 基础接口 =====
export interface BaseEntity {
  id: string;
  created_at: string;
  updated_at: string;
}

// ===== 看板接口 =====
export interface Board extends BaseEntity {
  user_id: string;
  name: string;
  description?: string;
  background_color: string;
  background_image?: string;
  is_archived: boolean;
  position: number;
  lists?: List[];
}

export interface BoardCreate {
  name: string;
  description?: string;
  background_color?: string;
  background_image?: string;
}

export interface BoardUpdate {
  name?: string;
  description?: string;
  background_color?: string;
  background_image?: string;
  is_archived?: boolean;
}

// ===== 列表接口 =====
export interface List extends BaseEntity {
  board_id: string;
  name: string;
  description?: string;
  position: number;
  is_archived: boolean;
  cards?: Card[];
}

export interface ListCreate {
  board_id: string;
  name: string;
  description?: string;
  position?: number;
}

export interface ListUpdate {
  name?: string;
  description?: string;
  position?: number;
  is_archived?: boolean;
}

// ===== 卡片接口 =====
export interface Card extends BaseEntity {
  list_id: string;
  title: string;
  description?: string;
  position: number;
  due_date?: string;
  priority: PriorityLevel;
  is_completed: boolean;
  is_archived: boolean;
  color?: string;
  tags: string[];
  completed_at?: string;
}

export interface CardCreate {
  list_id: string;
  title: string;
  description?: string;
  position?: number;
  due_date?: string;
  priority?: PriorityLevel;
  color?: string;
  tags?: string[];
}

export interface CardUpdate {
  title?: string;
  description?: string;
  position?: number;
  due_date?: string;
  priority?: PriorityLevel;
  is_completed?: boolean;
  color?: string;
  tags?: string[];
}

// ===== 拖拽操作接口 =====
export interface CardMoveRequest {
  card_id: string;
  source_list_id: string;
  target_list_id: string;
  new_position: number;
}

export interface DragState {
  draggedCard: Card | null;
  draggedList: List | null;
  dragOverListId: string | null;
  isDragging: boolean;
}

// ===== 批量操作接口 =====
export interface BulkCardUpdate {
  card_ids: string[];
  updates: CardUpdate;
}

export interface BulkCardMove {
  card_ids: string[];
  target_list_id: string;
}

// ===== 搜索和过滤接口 =====
export interface CardFilter {
  list_ids?: string[];
  priorities?: PriorityLevel[];
  is_completed?: boolean;
  has_due_date?: boolean;
  is_overdue?: boolean;
  tags?: string[];
}

export interface SearchRequest {
  query: string;
  board_id?: string;
  include_archived?: boolean;
}

export interface SearchResponse {
  cards: Card[];
  total_count: number;
  query: string;
}

// ===== 看板统计接口 =====
export interface BoardStats {
  total_cards: number;
  completed_cards: number;
  overdue_cards: number;
  cards_by_priority: Record<PriorityLevel, number>;
  cards_by_list: Record<string, number>;
}

// ===== 组件属性接口 =====
export interface BoardCardProps {
  board: Board;
  onClick: (board: Board) => void;
  onEdit: (board: Board) => void;
  onDelete: (boardId: string) => void;
}

export interface ListColumnProps {
  list: List;
  cards: Card[];
  onCardClick: (card: Card) => void;
  onCardCreate: (listId: string) => void;
  onListEdit: (list: List) => void;
  onListDelete: (listId: string) => void;
  onCardMove: (cardId: string, targetListId: string, newPosition: number) => void;
}

export interface CardItemProps {
  card: Card;
  listId: string;
  onClick: (card: Card) => void;
  onEdit: (card: Card) => void;
  onDelete: (cardId: string) => void;
  isDragging?: boolean;
}

export interface CardModalProps {
  card: Card | null;
  isOpen: boolean;
  onClose: () => void;
  onSave: (cardData: CardUpdate) => void;
  onDelete: (cardId: string) => void;
}

export interface BoardCreateModalProps {
  isOpen: boolean;
  onClose: () => void;
  onCreate: (boardData: BoardCreate) => void;
}

// ===== 拖拽相关接口 =====
export interface DragItem {
  id: string;
  type: 'card' | 'list';
  data: Card | List;
}

export interface DropResult {
  destinationId: string;
  position: number;
  type: 'card' | 'list';
}

export interface DragEndEvent {
  active: {
    id: string;
    data: {
      current: {
        type: string;
        card?: Card;
        list?: List;
      };
    };
  };
  over: {
    id: string;
    data: {
      current: {
        type: string;
        listId?: string;
        position?: number;
      };
    };
  } | null;
}

// ===== 状态管理接口 =====
export interface KanbanFilterState {
  searchQuery: string;
  priority: PriorityLevel | 'all';
  completed: boolean | null;
}

export interface KanbanDragState {
  isDragging: boolean;
  draggedCard: Card | null;
  draggedList: List | null;
  dragOverListId: string | null;
}

// ===== API响应接口 =====
export interface ApiResponse<T> {
  success: boolean;
  data: T;
  message?: string;
  error_code?: string;
  timestamp: string;
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

// ===== 错误处理接口 =====
export interface KanbanError {
  code: string;
  message: string;
  details?: any;
}

export interface ValidationError {
  field: string;
  message: string;
  value?: any;
}

// ===== 工具类型 =====
export type CardSortField = 'title' | 'due_date' | 'priority' | 'created_at' | 'updated_at';
export type CardSortOrder = 'asc' | 'desc';

export interface CardSortOptions {
  field: CardSortField;
  order: CardSortOrder;
}

export interface KanbanViewOptions {
  showCompleted: boolean;
  groupByPriority: boolean;
  compactMode: boolean;
  showTags: boolean;
  sortOptions: CardSortOptions;
}

// ===== 扩展类型 =====
export interface CardWithList extends Card {
  list_name?: string;
  board_name?: string;
}

export interface ListWithStats extends List {
  total_cards: number;
  completed_cards: number;
  overdue_cards: number;
}

export interface BoardWithStats extends Board {
  stats: BoardStats;
  lists_with_stats: ListWithStats[];
}

// ===== 表单接口 =====
export interface BoardFormData {
  name: string;
  description: string;
  background_color: string;
  background_image?: string;
}

export interface CardFormData {
  title: string;
  description: string;
  due_date?: string;
  priority: PriorityLevel;
  color?: string;
  tags: string[];
}

export interface ListFormData {
  name: string;
  description: string;
}

// ===== 事件处理接口 =====
export interface CardEventHandlers {
  onCardCreate: (listId: string, cardData: CardCreate) => void;
  onCardUpdate: (cardId: string, cardData: CardUpdate) => void;
  onCardDelete: (cardId: string) => void;
  onCardMove: (cardId: string, targetListId: string, newPosition: number) => void;
  onCardClick: (card: Card) => void;
}

export interface ListEventHandlers {
  onListCreate: (boardId: string, listData: ListCreate) => void;
  onListUpdate: (listId: string, listData: ListUpdate) => void;
  onListDelete: (listId: string) => void;
  onListReorder: (boardId: string, listIds: string[]) => void;
}

export interface BoardEventHandlers {
  onBoardCreate: (boardData: BoardCreate) => void;
  onBoardUpdate: (boardId: string, boardData: BoardUpdate) => void;
  onBoardDelete: (boardId: string) => void;
  onBoardSelect: (boardId: string) => void;
}